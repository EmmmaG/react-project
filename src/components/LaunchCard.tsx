import * as React from 'react'
import { withStyles, WithStyles, Theme } from '@material-ui/core/styles'
import Typography from '@material-ui/core/Typography'

import Button from '../components/Button'
import { Launch, Launchpad } from '../views/LaunchDataInterfaces'
import getNumberWithSuffix from '../utils/getNumberWithSuffix'
import getTime from '../utils/getTime'

const styles = (theme: Theme) => ({
	cardContainer: {
		display: 'flex',
		margin: '20px',
		justifyContent: 'start',
		alignItems: 'flex-start',
		borderBottom: `1px solid ${theme.palette.grey[200]}`,
	},
	picture: {
		height: '70px',
		width: '70px',
		marginRight: '10px',
	},
	textWrapper: {
		textAlign: 'left' as 'left',
	},
	failedText: {
		color: 'red',
	},
	subtitle: {
		fontSize: '12px',
		marginTop: '8px',
		color: theme.palette.grey[600],
	},
	buttonWrapper: {
		marginTop: '30px',
	},
	detailContainer: {
		marginTop: '10px',
		marginLeft: '10px',
	},
	flightNumberContainer: {
		marginTop: '15px',
		marginLeft: 'auto',
		textAlign: 'center' as 'center',
	},
})

const monthNames = [
	'January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December',
]

export interface Props extends WithStyles<typeof styles> {
	launchData: Launch
	launchpadsData: Launchpad[]
}

export class LaunchCard extends React.Component<Props> {
	private get isFailedMission() {
		const { launchData } = this.props

		return (!launchData.land_success || !launchData.launch_success)
	}

	private get launchPadName() {
		const { launchData, launchpadsData } = this.props
		const launchpadData = launchpadsData.find(data => data.id === launchData.launch_site.site_id)

		return launchpadData && launchpadData.full_name
	}

	private get subtitle() {
		const { launchData, classes} = this.props
		const trimedDte = launchData.launch_date_local.slice(0, -6)
		const date = new Date(trimedDte)

		const year = date.getFullYear()
		const month = monthNames[date.getMonth()]
		const day = getNumberWithSuffix(date.getDate())
		const time = getTime(date)

		return (
			<div className={classes.textWrapper}>
				<Typography className={classes.subtitle}>Launched <b>{`${day} ${month} ${year}`}</b> at <b>{time}</b> from <b>{this.launchPadName}</b></Typography>
			</div>
		)
	}

	private renderButtons = () => {
		const { launchData, classes} = this.props

		return (
			<div className={classes.buttonWrapper}>
				{launchData.links.reddit_campaign && (
					<Button buttonContent='Reddit Campaign' link={launchData.links.reddit_campaign} />
				)}
				{launchData.links.reddit_launch && (
					<Button buttonContent='Reddit Launch' link={launchData.links.reddit_launch} />
				)}
				{launchData.links.reddit_recovery && (
					<Button buttonContent='Reddit Recovery' link={launchData.links.reddit_recovery} />
				)}
				{launchData.links.reddit_media && (
					<Button buttonContent='Reddit Media' link={launchData.links.reddit_media} />
				)}
				{launchData.links.presskit && (
					<Button buttonContent='Press Kit' link={launchData.links.presskit} />
				)}
				{launchData.links.article_link && (
					<Button buttonContent='Article' link={launchData.links.article_link} />
				)}
				{launchData.links.video_link && (
					<Button buttonContent='Watch Video' link={launchData.links.video_link} />
				)}
			</div>
		)
	}

	private renderLaunchDetails = () => {
		const { launchData, classes } = this.props

		return (
			<div className={classes.detailContainer}>
				<div className={classes.textWrapper}>
					<Typography><b>{`${launchData.rocket.rocket_name} - ${launchData.payloads[0].payload_id}`}</b>{this.isFailedMission && (<span className={classes.failedText}> - Failed Mission</span>)}</Typography>
				</div>
				{this.subtitle}
				{this.renderButtons()}
			</div>
		)
	}

	private renderFlightNumber = () => {
		const { launchData, classes } = this.props

		return (
			<div className={classes.flightNumberContainer}>
				<Typography style={{fontSize: '18px'}}>{`#${launchData.flight_number}`}</Typography>
				<Typography style={{fontSize: '12px', color: 'grey'}}>Flight Number</Typography>
			</div>
		)
	}

	public render() {
		const { classes, launchData } = this.props

		return (
			<div className={classes.cardContainer}>
				<img
					src={launchData.links && launchData.links.mission_patch}
					className={classes.picture}
					alt='No Pic Available'
				/>
				{this.renderLaunchDetails()}
				{this.renderFlightNumber()}
			</div>
		)
	}
}

export default withStyles(styles)(LaunchCard)
