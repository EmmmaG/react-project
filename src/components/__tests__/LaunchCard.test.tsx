import * as React from 'react'
import { shallow } from 'enzyme'
import idObj from 'identity-obj-proxy'

import LaunchCard, { Props } from '../LaunchCard'

const launchData = [{
	cap_serial: null,
	core_serial: 'Merlin 1A',
	details: 'Engine failure at 33 seconds and loss of vehicle',
	flight_number: 1,
	landing_type: null,
	landing_vehicle: null,
	launch_date_local: '2006-03-25T10:30:00+12:00',
	launch_site: {
		site_id: 'kwajalein_atoll',
		site_name: 'Kwajalein Atoll',
	},
	links: {
		article_link: 'https://www.space.com/2196-spacex-inaugural-falcon-1-rocket-lost-launch.html',
		mission_patch: 'http://spacexpatchlist.space/images/thumbs/falcon_1_flight_1.png',
		video_link: 'https://www.youtube.com/watch?v=0a_00nJ_Y88',
	},
	payloads: [{
		customers: ['DARPA'],
		orbit: 'LEO',
		payload_id: 'FalconSAT-2',
		payload_mass_kg: 20,
		payload_mass_lbs: 43,
		payload_type: 'Satellite',
	}],
	rocket: {
		rocket_id: 'falcon1',
		rocket_name: 'Falcon 1',
		rocket_type: 'Merlin A',
	},
	telemetry: {
		flight_club: null,
	},
}] as any

export const launchpadsData = [{
	details: 'SpaceX original launch site, where all of the Falcon 1 launches occured. Abandoned as SpaceX decided against upgrading the pad to support Falcon 9.',
	full_name: 'Kwajalein Atoll',
	id: 'kwajalein_atoll',
	location: {name: 'Omelek Island', region: 'Marshall Islands', latitude: 9.0477206, longitude: 167.7431292},
	status: 'retired',
	vehicles_launched: 'falcon 1',
}] as any

export const props: Props = {
	launchData,
	launchpadsData,
	classes: idObj,
}

describe('LaunchCard', () => {
	it('renders correctly', () => {
		const wrapper = shallow(<LaunchCard {...props} />)
		expect(wrapper).toMatchSnapshot()
	})
})
