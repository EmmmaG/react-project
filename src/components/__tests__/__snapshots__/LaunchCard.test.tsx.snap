// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaunchCard renders correctly 1`] = `
<LaunchCard
  classes={
    Object {
      "buttonWrapper": "LaunchCard-buttonWrapper-6",
      "cardContainer": "LaunchCard-cardContainer-1",
      "detailContainer": "LaunchCard-detailContainer-7",
      "failedText": "LaunchCard-failedText-4",
      "flightNumberContainer": "LaunchCard-flightNumberContainer-8",
      "picture": "LaunchCard-picture-2",
      "subtitle": "LaunchCard-subtitle-5",
      "textWrapper": "LaunchCard-textWrapper-3",
    }
  }
  launchData={
    Array [
      Object {
        "cap_serial": null,
        "core_serial": "Merlin 1A",
        "details": "Engine failure at 33 seconds and loss of vehicle",
        "flight_number": 1,
        "landing_type": null,
        "landing_vehicle": null,
        "launch_date_local": "2006-03-25T10:30:00+12:00",
        "launch_site": Object {
          "site_id": "kwajalein_atoll",
          "site_name": "Kwajalein Atoll",
        },
        "links": Object {
          "article_link": "https://www.space.com/2196-spacex-inaugural-falcon-1-rocket-lost-launch.html",
          "mission_patch": "http://spacexpatchlist.space/images/thumbs/falcon_1_flight_1.png",
          "video_link": "https://www.youtube.com/watch?v=0a_00nJ_Y88",
        },
        "payloads": Array [
          Object {
            "customers": Array [
              "DARPA",
            ],
            "orbit": "LEO",
            "payload_id": "FalconSAT-2",
            "payload_mass_kg": 20,
            "payload_mass_lbs": 43,
            "payload_type": "Satellite",
          },
        ],
        "rocket": Object {
          "rocket_id": "falcon1",
          "rocket_name": "Falcon 1",
          "rocket_type": "Merlin A",
        },
        "telemetry": Object {
          "flight_club": null,
        },
      },
    ]
  }
  launchpadsData={
    Array [
      Object {
        "details": "SpaceX original launch site, where all of the Falcon 1 launches occured. Abandoned as SpaceX decided against upgrading the pad to support Falcon 9.",
        "full_name": "Kwajalein Atoll",
        "id": "kwajalein_atoll",
        "location": Object {
          "latitude": 9.0477206,
          "longitude": 167.7431292,
          "name": "Omelek Island",
          "region": "Marshall Islands",
        },
        "status": "retired",
        "vehicles_launched": "falcon 1",
      },
    ]
  }
/>
`;
